<a href="https://colab.research.google.com/github/ninaraima2/Projet_IA/blob/main/projet_IA_TRAORE_ADJA_MAMARO_NINA_RAIMA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Projet : IA appliquée pour la résolution d'un problème concret
**Cours** : Intelligence artificielle (INF38215-MS)  
**Nom et Prenom** : Adja Mamaro Nina Raima Traoré   
**Date de remise** : 14 Décembre 2025  
**Langage utilisé** : Python (Google Colab)  
**Contribution** :  
- Adja Mamaro Nina Raima Traoré : 100%


## 1. Contexte
Ce projet utilise le célèbre jeu de données **"Titanic: Machine Learning from Disaster"**. Il contient des informations sur les passagers du Titanic (âge, classe, sexe, etc.)
* **Source des données :** [Kaggle Titanic Competition](https://www.kaggle.com/competitions/titanic/data)
* **Tâche :** Il s'agit d'un problème de **classification binaire**. L'objectif est de développer des modèles d'apprentissage automatique capables de prédire si un passager a survécu ou non au naufrage en fonction de ses caractéristiques.

## 2. Objectifs
Les objectifs de ce travail sont doubles, théoriques et pratiques:

**Objectifs Théoriques :**
* Approfondir les connaissances en IA, spécifiquement en **Apprentissage d'Ensemble** (Ensemble Learning) et en **Apprentissage Profond** (Deep Learning).

**Objectifs Pratiques :**
* Développer des compétences pratiques à travers un pipeline complet.
* Effectuer la visualisation et le prétraitement des données (EDA).
* Entraîner et évaluer des modèles (XGBoost, Random Forest, Deep Learning)
* Optimiser les **hyperparamètres**.
* Analyser les **métriques d'évaluation** et interpréter les résultats.

## 3. Plan du projet
Le projet suit la structure suivante:
1.  **Partie Théorique :** Présentation des métriques, des courbes d'analyse et des algorithmes utilisés (XGBoost, Random Forest, Deep Learning).
2.  **Partie Pratique :**
    * Exploration et prétraitement des données (EDA).
    * Implémentation et optimisation du modèle XGBoost.
    * Implémentation et optimisation du modèle Random Forest.
    * Implémentation et optimisation du modèle d'Apprentissage Profond.
    * Comparaison finale des trois techniques.

---


# **Partie 1 : THEORIE**

## **1) Métriques pour la classification**

### **1.1 Accuracy (Exactitude)**
L'accuracy est la proportion de prédictions correctes sur l'ensemble des prédictions. Elle indique globalement à quel point un modèle de classification réussit à identifier correctement les données.

- **Avantages :**
  - Facile à calculer et à comprendre  
  - Donne une vision rapide de la performance globale  
  - Pertinente quand les classes sont bien équilibrées  

- **Limitations :**
  - Trompeuse si les classes sont déséquilibrées (un modèle peut sembler performant en ignorant la classe minoritaire)  
  - Ne distingue pas les types d'erreurs (faux positifs vs faux négatifs)  
  - Nécessite d'être complétée par d'autres métriques comme Precision, Recall et F1-score  

---

### **1.2 Precision (Précision)**
La precision mesure la proportion de prédictions positives qui sont réellement correctes. Autrement dit, parmi tous les cas que le modèle a classés comme « positifs », combien sont effectivement des vrais positifs.

- **Avantages :**
  - Utile quand les faux positifs sont coûteux ou dangereux  
  - Donne une idée claire de la fiabilité des prédictions positives  
  - Pertinente dans les contextes où l'on veut éviter de sur prédire une classe  

- **Limitations :**
  - Ignore les faux négatifs (ne dit rien sur les cas positifs ratés)  
  - Peut être trompeuse si utilisée seule  
  - Doit être complétée par le Recall et le F1-score  

---

### **1.3 Recall (Sensibilité/Rappel)**
Le recall mesure la proportion de vrais positifs correctement détectés par le modèle parmi tous les cas positifs existants. Il indique la capacité du modèle à ne pas rater les éléments importants d'une classe.

- **Avantages :**
  - Indispensable lorsque le coût d'un faux négatif est critique  
  - Permet d'évaluer la capacité du modèle à couvrir l'ensemble de la classe cible  

- **Limitations :**
  - Peut être artificiellement gonflé en prédisant simplement la classe positive pour tous les échantillons  
  - Si on optimise uniquement le rappel, la précision a tendance à diminuer

---

### **1.4 F1-Score**
Le F1-score est la moyenne harmonique entre la Precision et le Recall. Il combine ces deux métriques pour donner une mesure unique qui équilibre la fiabilité des prédictions positives et la capacité à retrouver tous les cas positifs.

- **Avantages :**
  - Offre une vision plus complète que l'accuracy seule  
  - Équilibre entre Precision et Recall, utile quand les classes sont déséquilibrées  
  - Pertinent dans les contextes où il faut à la fois éviter les faux positifs et ne pas rater de vrais positifs  

- **Limitations :**
  - Peut masquer un déséquilibre (si la précision est très bonne mais le rappel très faible, ou inversement)  
  - Moins intuitif à interpréter que l'accuracy  
  - Ne prend pas en compte la distribution des classes, dépend du contexte  

## **2) Courbes d'Évaluation Visuelle**
### 2.1 Courbes d’apprentissage (Learning Curves)
- **Utilisation :** Montrent l’évolution de la performance du modèle (accuracy ou erreur) en fonction du nombre d’exemples d’entraînement.  
- **Pourquoi utilisées :** Détecter le surapprentissage (*overfitting*) ou le sous-apprentissage (*underfitting*).  
- **Interprétation :**  
  - Courbe entraînement haute mais validation basse → surapprentissage.  
  - Les deux courbes basses → sous-apprentissage.  
  - Courbes qui convergent → bonne généralisation.  

---

### 2.2 Courbes ROC (Receiver Operating Characteristic)
- **Utilisation :** Représentent le compromis entre taux de vrais positifs (sensibilité) et taux de faux positifs.  
- **Pourquoi utilisées :** Comparer différents modèles ou seuils de décision.  
- **Interprétation :**  
  - Courbe proche du coin supérieur gauche → modèle performant.  
  - Aire sous la courbe (AUC) → indicateur global de performance.  

---

### 2.3 Courbes Precision-Recall
- **Utilisation :** Montrent le compromis entre précision et rappel selon différents seuils.  
- **Pourquoi utilisées :** Particulièrement adaptées aux données déséquilibrées.  
- **Interprétation :**  
  - Courbe proche du coin supérieur droit → bon équilibre précision/rappel.  
  - Permet de choisir le seuil optimal selon le contexte.  

---

### 2.4 Matrice de Confusion (Confusion Matrix)
- **Utilisation :** Tableau comparant les prédictions du modèle avec les vraies classes.  
- **Pourquoi utilisée :** Donne une vision détaillée des erreurs par classe.  
- **Interprétation :**  
  - Diagonale → prédictions correctes.  
  - Hors diagonale → erreurs (faux positifs, faux négatifs).  

---

### 2.5 Erreurs de Validation Croisée (Cross-Validation Errors)
- **Utilisation :** Mesurent la performance du modèle sur plusieurs sous‑ensembles des données.  
- **Pourquoi utilisées :** Évaluer la robustesse et la stabilité du modèle.  
- **Interprétation :**  
  - Erreurs similaires entre plis → modèle stable.  
  - Variations fortes → modèle sensible aux données.  

---

### 2.6 Erreurs d'Apprentissage (Learning Errors)
- **Utilisation :** Suivi de l'évolution de l'erreur au fil des itérations ou époques d'entraînement.  
- **Pourquoi utilisées :** Vérifier la convergence du modèle et ajuster les hyperparamètres.  
- **Interprétation :**  
  - Erreur entraînement ↓ mais validation ↑ → surapprentissage.  
  - Les deux erreurs restent élevées → sous-apprentissage.  
  - Les deux erreurs diminuent et se stabilisent → modèle bien entraîné.  

## **3)Techniques d'Apprentissage**
## 3.1 Ensemble Learning et XGBoost

### Présentation détaillée de l'algorithme XGBoost
XGBoost (Extreme Gradient Boosting) est une bibliothèque optimisée de Gradient Boosting, conçue pour être rapide, efficace et performante, notamment sur des ensembles de données volumineux.

### Origine
- Amélioration du Gradient Boosting classique.  
- Introduit des techniques de régularisation et d’optimisation pour réduire le surapprentissage (overfitting).  


### Fonctionnement
- Basé sur des **arbres de décision** construits séquentiellement.  
- Chaque nouvel arbre corrige les erreurs des précédents.  
- Intègre la **régularisation** (L1 et L2).  
- Exploite la **parallélisation** pour accélérer l’entraînement.  
- Gère automatiquement les données manquantes.

### Hyperparamètres clés et leur influence
- **n_estimators** : nombre d’arbres → plus élevé = meilleure performance mais risque de surapprentissage.  
- **max_depth** : profondeur des arbres → contrôle la complexité.  
- **learning_rate (eta)** : taux d’apprentissage → petit = apprentissage plus lent mais plus précis.  
- **subsample** : proportion d’échantillons utilisés → réduit le surapprentissage.   

### Avantages et utilisation
- Performance élevée et rapidité.  
- Flexibilité (classification, régression, ranking).  
- Robustesse grâce à la régularisation et gestion des données manquantes.  
- Très utilisé en recherche et en industrie.  

---

## 3.2 Random Forests

### Présentation détaillée de l'algorithme Random Forest
Random Forest est une méthode d’**ensemble learning** qui combine plusieurs arbres de décision pour améliorer la robustesse et la précision des prédictions.

### Origine
- Développé comme une extension des arbres de décision.  
- Inspiré du concept de **bagging** (Bootstrap Aggregating).  

### Fonctionnement
- Crée plusieurs arbres de décision indépendants à partir d’échantillons aléatoires des données.  
- Chaque arbre vote pour une classe, et la prédiction finale est obtenue par majorité.  
- Réduit la variance et améliore la stabilité par rapport à un seul arbre.  

### Hyperparamètres clés et leur influence
- **n_estimators** : nombre d’arbres → plus élevé = meilleure stabilité.  
- **max_depth** : profondeur des arbres → contrôle la complexité.  
- **max_features** : nombre de variables utilisées par arbre → équilibre entre diversité et performance.  
- **min_samples_split / min_samples_leaf** : taille minimale des sous‑ensembles → influence la granularité des arbres.  

### Avantages et utilisation
- Moins sensible au surapprentissage qu’un arbre unique.  
- Bonne performance sur des données variées.  
- Facile à utiliser et interpréter (importance des variables).  
- Utilisé en classification, régression et détection d’anomalies.  

---

## 3.3 Apprentissage profond (Deep Learning)

### Fonctionnement
L’apprentissage profond repose sur des **réseaux de neurones artificiels** organisés en couches. Chaque couche transforme les données et apprend des représentations de plus en plus abstraites.

### Architectures
- **Réseaux denses (Fully Connected)** : chaque neurone est connecté à tous les neurones de la couche suivante.  
- **Réseaux convolutifs (CNN)** : adaptés aux images, exploitent les filtres pour détecter des motifs locaux.  
- **Réseaux récurrents (RNN, LSTM, GRU)** : adaptés aux données séquentielles (texte, séries temporelles).  

### Hyperparamètres clés et leur influence
- **nombre de couches et de neurones** : contrôle la capacité d’apprentissage.  
- **learning_rate** : taux d’apprentissage → trop grand = instable, trop petit = lent.  
- **batch_size** : taille des lots → influence la vitesse et la stabilité.  
- **epochs** : nombre de passes sur les données → trop élevé = surapprentissage.  
- **dropout** : régularisation pour éviter le surapprentissage.  

### Avantages et utilisation
- Excellente performance sur des données complexes (images, texte, audio).  
- Capacité à apprendre des représentations automatiquement.  
- Utilisé dans la vision par ordinateur, le traitement du langage naturel, la reconnaissance vocale, etc.  
- Très flexible mais demande beaucoup de données et de puissance de calcul.  

---


# **Partie 2: PRATIQUE**

Implementation des trois techniques sur le dataset titanic

Ce script contient :

1. **ÉTAPE 01** : Exploration des Données (EDA)
   - Chargement de train.csv
   - Statistiques descriptives
   - Histogrammes, corrélations, équilibre des classes
   - Traitement des valeurs manquantes
   - Visualisations : histogrammes.png, correlation_matrix.png, boxplots_by_target.png, pairplot_top_features.png

2. **ÉTAPE 02** : XGBoost
   - GridSearchCV pour tuner hyperparamètres
   - 5-Fold Stratified Cross-Validation
   - Entraînement sur tout train.csv

3. **ÉTAPE 03** : Random Forest
   - GridSearchCV pour tuner hyperparamètres
   - 5-Fold Stratified Cross-Validation
   - Entraînement sur tout train.csv

4. **ÉTAPE 04** : Deep Learning (MLP)
   - GridSearchCV pour tuner hyperparamètres
   - 5-Fold Stratified Cross-Validation
   - Entraînement sur tout train.csv

5. **ÉTAPE 05** : Prédictions et Soumissions
   - Chargement de test.csv
   - Prédictions avec les 3 modèles
   - Génération des fichiers de soumission :
     - submission_xgboost.csv
     - submission_randomforest.csv
     - submission_neural_network.csv

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

# Importer les outils de prétraitement et d'évaluation
from sklearn.model_selection import train_test_split, GridSearchCV, StratifiedKFold
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, roc_curve, roc_auc_score
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.impute import SimpleImputer

# Importer les modèles d'apprentissage
from xgboost import XGBClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neural_network import MLPClassifier

# ==============================================================================
# ÉTAPE 01: EXPLORATION DES DONNÉES (EDA)
# ==============================================================================
# Objectif: Analyser les données pour comprendre leur structure et identifier
# les problèmes potentiels (valeurs manquantes, déséquilibre des classes, etc.)
# ==============================================================================

print("="*80)
print("ÉTAPE 01: EXPLORATION DES DONNÉES (EDA)")
print("="*80)

# Charger le fichier CSV contenant les données d'entraînement
train_data = pd.read_csv("train.csv")

# Afficher des informations basiques sur le dataset
print("\n1. APERÇU GÉNÉRALE DES DONNÉES")
print("-" * 80)
print(f"Dimensions du dataset: {train_data.shape}")
print("\nPremières lignes du dataset:")
print(train_data.head())

# Informations générales
print("\n2. INFORMATIONS GÉNÉRALES")
print("-" * 80)
print(train_data.info())

# Statistiques descriptives
print("\n3. STATISTIQUES DESCRIPTIVES")
print("-" * 80)
# Calculer les statistiques descriptives (moyenne, médiane, écart-type, etc.)
print(train_data.describe())

# Valeurs manquantes
print("\n4. VALEURS MANQUANTES")
print("-" * 80)
# Identifier les colonnes avec des valeurs manquantes
missing_values = train_data.isnull().sum()
missing_percent = (missing_values / len(train_data)) * 100
missing_df = pd.DataFrame({'Manquantes': missing_values, 'Pourcentage': missing_percent})
print(missing_df[missing_df['Manquantes'] > 0])
if missing_df['Manquantes'].sum() == 0:
    print("Aucune valeur manquante détectée.")


# Identifier la colonne cible et les caractéristiques
print("\n5. STRUCTURE DES DONNÉES")
print("-" * 80)
print(f"Colonnes: {list(train_data.columns)}")

# Identifier la colonne cible 'Survived'
target_column = 'Survived'
print(f"Colonne cible : {target_column}")

# Analyser l'équilibre des classes
print("\n6. DISTRIBUTION DES CLASSES")
print("-" * 80)
class_distribution = train_data[target_column].value_counts()
print(f"\nDistribution des classes:")
print(class_distribution)
print(f"\nProportions:")
print(train_data[target_column].value_counts(normalize=True) * 100)



# Identifier les colonnes numériques pour les visualisations
numeric_columns = train_data.select_dtypes(include=[np.number]).columns.tolist()
# Exclure PassengerId et autres identifiants qui n'apportent pas d'info analytique
columns_to_exclude = ['PassengerId', target_column ]
numeric_columns = [col for col in numeric_columns if col not in columns_to_exclude]
print("\nColonnes numériques sélectionnées pour visualisation:", numeric_columns)

# Créer les histogrammes
plt.figure(figsize=(15, 12))

n_cols = len(numeric_columns)
n_rows = (n_cols + 2) // 3

for i, col in enumerate(numeric_columns, 1):
    plt.subplot(n_rows, 3, i)
    plt.hist(train_data[col].dropna(), bins=30, edgecolor='black', alpha=0.7)
    plt.title(f'Distribution de {col}')
    plt.xlabel(col)
    plt.ylabel('Fréquence')

plt.tight_layout()
plt.savefig('histogrammes.png', dpi=100, bbox_inches='tight')
print("\nGraphique sauvegardé: histogrammes.png")
plt.close()


# Matrice de corrélation
print("\n7. MATRICE DE CORRÉLATION")
print("-" * 80)
correlation_matrix = train_data[numeric_columns + [target_column]].corr()
print(correlation_matrix)

plt.figure(figsize=(10, 8))
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', center=0,
            fmt='.2f', cbar_kws={'label': 'Corrélation'})
plt.title('Matrice de Corrélation')
plt.tight_layout()
plt.savefig('correlation_matrix.png', dpi=100, bbox_inches='tight')
print("\nGraphique sauvegardé: correlation_matrix.png")
plt.close()


# Visualisations : comparer les variables selon la survie
# Les scatterplots avec l'index ne donnent pas d'info utile.
# À la place, on utilise des boxplots pour voir la différence
# de distribution des variables numériques entre survivants et non-survivants.
if len(numeric_columns) > 0:
    n_plots = len(numeric_columns)
    n_cols_bp = 4
    n_rows_bp = (n_plots + n_cols_bp - 1) // n_cols_bp
    plt.figure(figsize=(4 * n_cols_bp, 3 * max(2, n_rows_bp)))
    for i, col in enumerate(numeric_columns, 1):
        plt.subplot(n_rows_bp, n_cols_bp, i)
        sns.boxplot(x=train_data[target_column].astype(str), y=train_data[col])
        plt.title(f'Boxplot: {col} par classe')
        plt.xlabel('Classe')
        plt.ylabel(col)
    plt.tight_layout()
    plt.savefig('diagrammes_de_dispersion.png', dpi=100, bbox_inches='tight')
    print("Graphique sauvegardé: diagrammes_de_dispersiont.png")
    plt.close()


# Traiter les valeurs manquantes
print("\n8. TRAITEMENT DES VALEURS MANQUANTES")
print("-" * 80)
imputer = SimpleImputer(strategy='median')
numeric_data = train_data[numeric_columns].copy()
numeric_data_imputed = pd.DataFrame(
    imputer.fit_transform(numeric_data),
    columns=numeric_columns
)
train_data_processed = train_data.copy()
train_data_processed[numeric_columns] = numeric_data_imputed
print("Valeurs manquantes traitées avec la stratégie 'median'")

# Encoder les variables catégories
categorical_columns = train_data_processed.select_dtypes(include=['object']).columns.tolist()
if target_column in categorical_columns:
    categorical_columns.remove(target_column)

label_encoders = {}
for col in categorical_columns:
    le = LabelEncoder()
    train_data_processed[col] = le.fit_transform(train_data_processed[col].astype(str))
    label_encoders[col] = le

# Encoder la cible
if train_data_processed[target_column].dtype == 'object':
    le_target = LabelEncoder()
    train_data_processed[target_column] = le_target.fit_transform(train_data_processed[target_column])

print("\nEDA terminée!")
print("="*80)


# Préparer les données pour le modèle
X = train_data_processed.drop(columns=[target_column])
y = train_data_processed[target_column]

# Normaliser les caractéristiques
scaler = StandardScaler()
X_scaled = pd.DataFrame(
    scaler.fit_transform(X),
    columns=X.columns
)

# Utiliser TOUT train.csv pour l'entraînement (pas de division train/test)
# On évaluera sur test.csv après
X_train, y_train = X_scaled, y

print(f"\nDimensions d'entraînement: {X_train.shape}")
print("Note: Tout le dataset train.csv est utilisé pour l'entraînement.")


# Fonction pour évaluer les modèles avec validation croisée et générer les courbes ROC
def evaluate_model_cv(model, X, y, cv, model_name, color='blue'):
    """
    Évalue un modèle avec validation croisée et génère les courbes ROC et métriques
    """
    from sklearn.metrics import roc_curve, auc

    print(f"\n4. MÉTRIQUES ET COURBES ROC")
    print("-" * 80)

    y_pred_list = []
    y_pred_proba_list = []
    y_true_list = []

    tprs = []
    aucs = []
    mean_fpr = np.linspace(0, 1, 100)

    fold = 1
    for train_idx, val_idx in cv.split(X, y):
        X_train_fold, X_val_fold = X.iloc[train_idx], X.iloc[val_idx]
        y_train_fold, y_val_fold = y.iloc[train_idx], y.iloc[val_idx]

        # Entraîner le modèle sur le fold
        model.fit(X_train_fold, y_train_fold)

        # Prédictions
        y_pred_fold = model.predict(X_val_fold)
        y_pred_proba_fold = model.predict_proba(X_val_fold)[:, 1]

        y_pred_list.append(y_pred_fold)
        y_pred_proba_list.append(y_pred_proba_fold)
        y_true_list.append(y_val_fold.values)

        # Calculer ROC
        fpr, tpr, _ = roc_curve(y_val_fold, y_pred_proba_fold)
        roc_auc = auc(fpr, tpr)
        aucs.append(roc_auc)

        # Interpoler la courbe ROC
        tpr_interp = np.interp(mean_fpr, fpr, tpr)
        tpr_interp[0] = 0.0
        tprs.append(tpr_interp)

        acc = accuracy_score(y_val_fold, y_pred_fold)
        print(f"Fold {fold}: Accuracy={acc:.4f}, ROC-AUC={roc_auc:.4f}")
        fold += 1

    # Métriques moyennes
    y_pred_all = np.concatenate(y_pred_list)
    y_pred_proba_all = np.concatenate(y_pred_proba_list)
    y_true_all = np.concatenate(y_true_list)

    acc_mean = accuracy_score(y_true_all, y_pred_all)
    auc_mean = np.mean(aucs)
    auc_std = np.std(aucs)

    print(f"\nRésultats moyens (5-fold):")
    print(f"  Accuracy: {acc_mean:.4f}")
    print(f"  ROC-AUC: {auc_mean:.4f} (+/- {auc_std:.4f})")
    print(f"\nRapport de classification:")
    print(classification_report(y_true_all, y_pred_all))

    # Matrice de confusion
    cm = confusion_matrix(y_true_all, y_pred_all)
    plt.figure(figsize=(8, 6))
    sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', cbar_kws={'label': 'Nombre'})
    plt.title(f'Matrice de Confusion - {model_name} (5-fold CV)')
    plt.ylabel('Vraie valeur')
    plt.xlabel('Prédiction')
    plt.tight_layout()
    plt.savefig(f'{model_name.lower()}_confusion_matrix_cv.png', dpi=100, bbox_inches='tight')
    plt.close()
    print(f"Graphique sauvegardé: {model_name.lower()}_confusion_matrix_cv.png")

    # Courbe ROC moyenne
    mean_tpr = np.mean(tprs, axis=0)
    mean_tpr[-1] = 1.0

    plt.figure(figsize=(8, 6))
    plt.plot(mean_fpr, mean_tpr, color=color, lw=2,
             label=f'{model_name} (AUC = {auc_mean:.4f} +/- {auc_std:.4f})')
    plt.fill_between(mean_fpr,
                     np.min(tprs, axis=0),
                     np.max(tprs, axis=0),
                     color=color, alpha=0.2)
    plt.plot([0, 1], [0, 1], 'k--', label='Aléatoire')
    plt.xlabel('Taux de faux positifs')
    plt.ylabel('Taux de vrais positifs')
    plt.title(f'Courbe ROC - {model_name} (5-fold CV)')
    plt.legend(loc='lower right')
    plt.grid()
    plt.tight_layout()
    plt.savefig(f'{model_name.lower()}_roc_curve_cv.png', dpi=100, bbox_inches='tight')
    plt.close()
    print(f"Graphique sauvegardé: {model_name.lower()}_roc_curve_cv.png")

    return auc_mean, acc_mean

# Stocker les résultats pour comparaison
results = {}

# ========================
# ÉTAPE 02: APPRENTISSAGE AVEC XGBOOST
# ========================
print("\n" + "="*80)
print("ÉTAPE 02: APPRENTISSAGE AVEC XGBOOST")
print("="*80)

# Stratégie K-Fold pour la validation croisée
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# GridSearch pour XGBoost
print("\n1. RECHERCHE DES HYPERPARAMÈTRES AVEC GRIDSEARCH ET K-FOLD")
print("-" * 80)

param_grid_xgb = {
    'n_estimators': [50, 100, 150],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2],
    'subsample': [0.8, 0.9, 1.0]
}

xgb_model = XGBClassifier(random_state=42, verbose=0)
grid_search_xgb = GridSearchCV(
    xgb_model, param_grid_xgb, cv=skf, scoring='roc_auc', n_jobs=-1, verbose=1
)
grid_search_xgb.fit(X_train, y_train)

print(f"\nMeilleurs hyperparamètres XGBoost: {grid_search_xgb.best_params_}")
print(f"Meilleur score CV (ROC-AUC): {grid_search_xgb.best_score_:.4f}")

# Entraîner le modèle final sur TOUT le dataset avec les meilleurs hyperparamètres
best_xgb = grid_search_xgb.best_estimator_
print("\n2. ENTRAÎNEMENT DU MODÈLE FINAL")
print("-" * 80)
best_xgb.fit(X_train, y_train)
print("Le modèle XGBoost a été entraîné sur tout le dataset train.csv")
print("avec les hyperparamètres optimisés par GridSearchCV (5-fold).")

# Évaluer avec validation croisée et générer les courbes
print("\n3. ÉVALUATION EN VALIDATION CROISÉE (5-FOLD)")
print("-" * 80)
xgb_for_eval = XGBClassifier(**grid_search_xgb.best_params_, random_state=42, verbose=0)
auc_xgb, acc_xgb = evaluate_model_cv(xgb_for_eval, X_train, y_train, skf, 'XGBoost', color='blue')
results['XGBoost'] = {'AUC': auc_xgb, 'Accuracy': acc_xgb}

# ==============================================================================
# ÉTAPE 03: APPRENTISSAGE AVEC RANDOM FORESTS
# ==============================================================================
# Objectif: Entraîner un modèle Random Forest avec GridSearch et k-fold
# ==============================================================================

print("\n" + "="*80)
print("ÉTAPE 03: APPRENTISSAGE AVEC RANDOM FORESTS")
print("="*80)

print("\n1. RECHERCHE DES HYPERPARAMÈTRES AVEC GRIDSEARCH ET K-FOLD")
print("-" * 80)

param_grid_rf = {
    'n_estimators': [50, 100, 150],
    'max_depth': [10, 15, 20, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

rf_model = RandomForestClassifier(random_state=42)
grid_search_rf = GridSearchCV(
    rf_model, param_grid_rf, cv=skf, scoring='roc_auc', n_jobs=-1, verbose=1
)
grid_search_rf.fit(X_train, y_train)

print(f"\nMeilleurs hyperparamètres Random Forest: {grid_search_rf.best_params_}")
print(f"Meilleur score CV (ROC-AUC): {grid_search_rf.best_score_:.4f}")

# Entraîner le modèle final sur TOUT le dataset avec les meilleurs hyperparamètres
best_rf = grid_search_rf.best_estimator_
print("\n2. ENTRAÎNEMENT DU MODÈLE FINAL")
print("-" * 80)
best_rf.fit(X_train, y_train)
print("Le modèle Random Forest a été entraîné sur tout le dataset train.csv")
print("avec les hyperparamètres optimisés par GridSearchCV (5-fold).")

# Évaluer avec validation croisée et générer les courbes
print("\n3. ÉVALUATION EN VALIDATION CROISÉE (5-FOLD)")
print("-" * 80)
rf_for_eval = RandomForestClassifier(**grid_search_rf.best_params_, random_state=42)
auc_rf, acc_rf = evaluate_model_cv(rf_for_eval, X_train, y_train, skf, 'RandomForest', color='green')
results['Random Forest'] = {'AUC': auc_rf, 'Accuracy': acc_rf}

# ==============================================================================
# ÉTAPE 04: APPRENTISSAGE PROFOND (RÉSEAU DE NEURONES)
# ==============================================================================
# Objectif: Entraîner un MLP (Multi-Layer Perceptron) avec GridSearch et k-fold
# ==============================================================================

print("\n" + "="*80)
print("ÉTAPE 04: APPRENTISSAGE PROFOND (RÉSEAU DE NEURONES)")
print("="*80)

print("\n1. RECHERCHE DES HYPERPARAMÈTRES AVEC GRIDSEARCH ET K-FOLD")
print("-" * 80)

param_grid_nn = {
    'hidden_layer_sizes': [(64, 32), (128, 64), (256, 128)],
    'activation': ['relu', 'tanh'],
    'alpha': [0.0001, 0.001, 0.01],
    'learning_rate_init': [0.001, 0.01]
}

nn_model = MLPClassifier(random_state=42, max_iter=500, early_stopping=True, validation_fraction=0.1)
grid_search_nn = GridSearchCV(
    nn_model, param_grid_nn, cv=skf, scoring='roc_auc', n_jobs=-1, verbose=1
)
grid_search_nn.fit(X_train, y_train)

print(f"\nMeilleurs hyperparamètres Neural Network: {grid_search_nn.best_params_}")
print(f"Meilleur score CV (ROC-AUC): {grid_search_nn.best_score_:.4f}")

# Entraîner le modèle final sur TOUT le dataset avec les meilleurs hyperparamètres
best_nn = grid_search_nn.best_estimator_
print("\n2. MODÈLE FINAL ENTRAÎNÉ")
print("-" * 80)
print("Le modèle Neural Network (MLP) a été entraîné sur tout le dataset train.csv")
print("avec les hyperparamètres optimisés par GridSearchCV (5-fold).")

# Évaluer avec validation croisée et générer les courbes
print("\n3. ÉVALUATION EN VALIDATION CROISÉE (5-FOLD)")
print("-" * 80)
nn_for_eval = MLPClassifier(**grid_search_nn.best_params_, random_state=42, max_iter=500, early_stopping=True, validation_fraction=0.1)
auc_nn, acc_nn = evaluate_model_cv(nn_for_eval, X_train, y_train, skf, 'NeuralNetwork', color='purple')
results['Neural Network'] = {'AUC': auc_nn, 'Accuracy': acc_nn}

# ==============================================================================
# ÉTAPE 05: PRÉDICTIONS SUR L'ENSEMBLE DE TEST
# ==============================================================================
# Objectif: Charger test.csv, prétraiter les données de la même façon,
# et générer des prédictions pour chaque passager
# ==============================================================================

print("\n" + "="*80)
print("ÉTAPE 05: PRÉDICTIONS SUR L'ENSEMBLE DE TEST")
print("="*80)

# Charger l'ensemble de test
test_data = pd.read_csv("test.csv")
print(f"\nDimensions de test.csv: {test_data.shape}")
print("Les 5 premières lignes de test.csv:")
print(test_data.head())

# Sauvegarder les PassengerId pour la soumission finale
passenger_ids = test_data['PassengerId'].copy()

# Appliquer le même prétraitement que sur train_data
# 1. Identifier les colonnes numériques (sauf PassengerId)
test_numeric_cols = test_data.select_dtypes(include=[np.number]).columns.tolist()
test_numeric_cols = [col for col in test_numeric_cols if col not in ['PassengerId']]

# 2. Traiter les valeurs manquantes avec l'imputer déjà entraîné
test_data_processed = test_data.copy()
test_numeric_data = test_data_processed[test_numeric_cols].copy()
test_numeric_data_imputed = pd.DataFrame(
    imputer.transform(test_numeric_data),
    columns=test_numeric_cols
)
test_data_processed[test_numeric_cols] = test_numeric_data_imputed

# 3. Encoder les variables catégorielles avec les encoders déjà entraînés
test_categorical_cols = test_data_processed.select_dtypes(include=['object']).columns.tolist()
test_categorical_cols = [col for col in test_categorical_cols if col not in ['PassengerId']]

for col in test_categorical_cols:
    if col in label_encoders:
        # Gérer les valeurs inconnues en les remplaçant par une valeur par défaut
        # ou en les codant avec un code spécial (-1 ou un index valide)
        encoder = label_encoders[col]
        test_values = test_data_processed[col].astype(str)

        # Remplacer les valeurs inconnues par la classe la plus fréquente (mode) du train
        known_mask = test_values.isin(encoder.classes_)
        if not known_mask.all():
            # Remplacer les valeurs inconnues par la première classe connue (ou mode)
            default_class = encoder.classes_[0]
            test_values[~known_mask] = default_class

        test_data_processed[col] = encoder.transform(test_values)
    else:
        # Si la colonne n'existe pas dans les encoders (rare), encoder manuellement
        le_temp = LabelEncoder()
        test_data_processed[col] = le_temp.fit_transform(test_data_processed[col].astype(str))

# 4. Sélectionner les mêmes features que dans X_train
X_test_final = test_data_processed[X.columns].copy()

# 5. Normaliser avec le scaler déjà entraîné
X_test_final_scaled = pd.DataFrame(
    scaler.transform(X_test_final),
    columns=X.columns
)

print("\nPrétraitement de test.csv terminé!")
print(f"Dimensions après prétraitement: {X_test_final_scaled.shape}")

# Faire des prédictions avec les trois meilleurs modèles
print("\n1. PRÉDICTIONS AVEC LES TROIS MODÈLES")
print("-" * 80)

test_pred_xgb = best_xgb.predict(X_test_final_scaled)
test_pred_rf = best_rf.predict(X_test_final_scaled)
test_pred_nn = best_nn.predict(X_test_final_scaled)

print(f"Prédictions XGBoost (premiers 10): {test_pred_xgb[:10]}")
print(f"Prédictions Random Forest (premiers 10): {test_pred_rf[:10]}")
print(f"Prédictions Neural Network (premiers 10): {test_pred_nn[:10]}")

# Créer les fichiers de soumission pour chaque modèle
print("\n2. GÉNÉRATION DES FICHIERS DE SOUMISSION")
print("-" * 80)

# Soumission XGBoost
submission_xgb = pd.DataFrame({
    'PassengerId': passenger_ids,
    'Survived': test_pred_xgb
})
submission_xgb.to_csv('submission_xgboost.csv', index=False)
print("Fichier sauvegardé: submission_xgboost.csv")

# Soumission Random Forest
submission_rf = pd.DataFrame({
    'PassengerId': passenger_ids,
    'Survived': test_pred_rf
})
submission_rf.to_csv('submission_randomforest.csv', index=False)
print("Fichier sauvegardé: submission_randomforest.csv")

# Soumission Neural Network
submission_nn = pd.DataFrame({
    'PassengerId': passenger_ids,
    'Survived': test_pred_nn
})
submission_nn.to_csv('submission_neural_network.csv', index=False)
print("Fichier sauvegardé: submission_neural_network.csv")

# Afficher un résumé des prédictions
print("\n4. RÉSUMÉ DES PRÉDICTIONS")
print("-" * 80)
summary_df = pd.DataFrame({
    'Modèle': ['XGBoost', 'Random Forest', 'Neural Network'],
    'Survivants prédits': [
        test_pred_xgb.sum(),
        test_pred_rf.sum(),
        test_pred_nn.sum()
    ],
    'Taux de survie (%)': [
        (test_pred_xgb.sum() / len(test_pred_xgb)) * 100,
        (test_pred_rf.sum() / len(test_pred_rf)) * 100,
        (test_pred_nn.sum() / len(test_pred_nn)) * 100
    ]
})
print(summary_df.to_string(index=False))

print("\n" + "="*80)
print("ANALYSE COMPLÈTE TERMINÉE!")
print("="*80)
print("\nFichiers de soumission générés:")
print("  - submission_xgboost.csv")
print("  - submission_randomforest.csv")
print("  - submission_neural_network.csv")

ÉTAPE 01: EXPLORATION DES DONNÉES (EDA)

1. APERÇU GÉNÉRALE DES DONNÉES
--------------------------------------------------------------------------------
Dimensions du dataset: (891, 12)

Premières lignes du dataset:
   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S 


# COMPARAISON ET INTERPRÉTATION DES RÉSULTATS

## Résultats et Classement

- **XGBoost** : Accuracy ( 84%), ROC-AUC ( 89%)

- **Random Forest** : Accuracy ( 81%),  ROC-AUC ( 86%)

- **Deep Learning (MLP)** : Accuracy ( 81%), ROC-AUC ( 86%)

## Analyse des modèles

- **XGBoost** est le plus performant. Sa méthode séquentielle corrige les erreurs pas à pas, ce qui lui donne un avantage sur petit dataset tabulaire.

- **Random** Forest reste stable et robuste, mais ses arbres indépendants n'optimisent pas les erreurs comme XGBoost, d'où des scores légèrement inférieurs.

 - **Neural Network**(MLP) est puissant en théorie, mais avec seulement 891 exemples il n'arrive pas à bien généraliser. Les réseaux profonds demandent beaucoup plus de données pour être efficaces.

## Pourquoi les résutats different ?

- **Architectures différentes** : XGBoost apprend en corrigeant ses erreurs pas à pas, ce qui le rend plus précis. Random Forest fait voter plusieurs arbres en parallèle, ce qui est stable mais moins optimisé. Le réseau de neurones ajuste ses poids par rétropropagation, c'est puissant mais ça demande beaucoup de données.

- **Taille du dataset**: Le dataset Titanic est petit (891 passagers) et tabulaire. XGBoost s'adapte bien à ce type de données et reste performant. Random Forest fonctionne correctement mais reste un peu en dessous. Le réseau de neurones n'a pas assez de données pour apprendre correctement, ce qui le rend instable.

## Conclusion

Dans ce projet, j'ai comparé trois modèles sur le dataset Titanic. Après l'EDA, le prétraitement et la validation croisée, j'ai constaté que** XGBoost est le plus performant** (Accuracy ≈ 84%, AUC ≈ 0.89). Random Forest est correct mais moins optimisé, et le réseau de neurones souffre du petit nombre d'exemples. Pour la soumission finale, je retiens donc **XGBoost**, car il combine performance, stabilité et adaptation aux données tabulaires.

# RÉFÉRENCES

## Données
- Kaggle Titanic Dataset: https://www.kaggle.com/competitions/titanic/data

## Bibliothèques Python
- Pandas: https://pandas.pydata.org/
- NumPy: https://numpy.org/
- Scikit-learn: https://scikit-learn.org/
- XGBoost: https://xgboost.readthedocs.io/
- Matplotlib: https://matplotlib.org/
- Seaborn: https://seaborn.pydata.org/
- TensorFlow/Keras: https://www.tensorflow.org/

## Théorie et Tutoriels
- Chen & Guestrin (2016). "XGBoost: A Scalable Tree Boosting System"
- Breiman, L. (2001). "Random Forests". Machine Learning, 45(1), 5-32
- Goodfellow, I., Bengio, Y., & Courville, A. (2016). "Deep Learning". MIT Press
- Scikit-learn Documentation: https://scikit-learn.org/stable/documentation.html
- Medium articles sur XGBoost et Random Forest
- Kaggle Kernels et discussions

## Concepts
- Cross-Validation: https://scikit-learn.org/stable/modules/cross_validation.html
- GridSearchCV: https://scikit-learn.org/stable/modules/grid_search.html
- ROC Curves: https://scikit-learn.org/stable/auto_examples/model_selection/plot_roc.html